File: 4 kyu/Matrix Determinant/solution.cs
using System;

public class Matrix
{

  public static int Determinant(int[][] matrix)
  {
    int det = 0;
    if (matrix.Length != matrix[0].Length)
      return -1;
    if (matrix.Length == 1)
      return matrix[0][0];

    for (int i = 0; i < matrix.Length; i++)
      det += (int)Math.Pow(-1, i) * matrix[0][i] * Determinant(Minor(matrix, i));

    return det;
  }

  public static int[][] Minor(int[][] matrix, int pos)
  {
    int[][] minor = new int[matrix.Length - 1][];
    for (int i = 0; i < minor.Length; i++)
      minor[i] = new int[minor.Length];

    for (int i = 1; i < matrix.Length; i++)
    {
      for (int j = 0; j < pos; j++)
        minor[i - 1][j] = matrix[i][j];
      for (int j = pos + 1; j < matrix.Length; j++)
        minor[i - 1][j - 1] = matrix[i][j];
    }
        return minor;
  }
}
_____________________________________________
using System;

public class Matrix
{
   public static int[][] Minor(int[][] matrix, int pos)
   {
      int[][] minor = new int[matrix.Length - 1][];
      for(int i = 0; i < minor.Length; i++)
      {
          minor[i] = new int[minor.Length];  
      }
       
      for(int i = 1; i < matrix.Length; i++)
      {
          for(int j = 0; j < pos; j++)
          {
              minor[i -1][j] = matrix[i][j];
          }
          for(int j = pos + 1; j < matrix.Length; j++)
          {
              minor[i - 1][j - 1] = matrix[i][j];
          }
      }
     
      return minor;
   }
  
   public static int Determinant(int[][] matrix)
   {
       int det = 0;
       if(matrix.Length != matrix[0].Length) return -1;
       if(matrix.Length == 1) return matrix[0][0];
     
       for(int i = 0; i < matrix.Length; i++)
       {
           det+= (int)Math.Pow(-1, i) * matrix[0][i] * Determinant(Minor(matrix, i));         
       }
     
       return det;
   }
}
_____________________________________________
using System;
using System.Collections.Generic;
using System.Linq;

public class Matrix
{
  public static int Determinant(int[][] matrix)
  {
    int sum = 0;
    
    if (matrix.Length == 1)
      return matrix[0][0];
    
    else
      for (int i = 0; i < matrix.Length; i++)
        sum += matrix[0][i] * (int)Math.Pow(-1, i) * Determinant(Minor(matrix, i));

    return sum;
  }
  
  private static int[][] Minor(int[][] matrix, int index)
  {
    int[][] minor = new int[matrix.Length - 1][];
    
    for (int i = 0; i < minor.Length; i++)
      minor[i] = new int[minor.Length];

    for (int i = 1; i < matrix.Length; i++)
    {
      for (int j = 0; j < index; j++)
        minor[i - 1][j] = matrix[i][j];
      for (int j = index + 1; j < matrix.Length; j++)
        minor[i - 1][j - 1] = matrix[i][j];
    }
    
    return minor;
  }
}
_____________________________________________
public class Matrix
{
    public static int Determinant(int[][] matrix)
    {
        if (matrix.Length == 1)
        {
            return matrix[0][0];
        }
        
        if (matrix.Length == 2)
        {
            return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0];
        }

        int res = 0;
        int sign = 1;
        for (int i = 0; i < matrix[0].Length; i++)
        {
            res += sign * matrix[0][i] * Determinant(Minor(matrix, i));
            sign *= -1;
        }

        return res;
    }

    public static int[][] Minor(int[][] matrix, int index)
    {
        var len = matrix.Length - 1;
        int[][] result = new int[len][];

        // Initialize
        for (int c = 0; c < len; c++)
        {
            result[c] = new int[len];
        }

        for (int i = 0; i < result.Length; i++)
        {
            int x = 0;
            for (int j = 0; j < result.Length; j++)
            {
                if (x == index)
                    x++;

                result[i][j] = matrix[i + 1][x];
                x++;
            }
        }
        return result;
    }
}
_____________________________________________
public class Matrix
{
   public static int Determinant(int[][] matrix)
   {
       if(matrix[0].Length == 1)
           return matrix[0][0];
       return DetHelper(matrix);
   }
   
   public static int DetHelper(int[][] matrix)
   {
       int l = matrix[0].Length;
       if(l == 2)
           return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0];
           
       int total = 0;
       for(int i = 0; i < l; i++)
       {
           if(i % 2 == 0)
               total += matrix[0][i] * DetHelper(RemoveColumn(matrix, i));
           else
               total -= matrix[0][i] * DetHelper(RemoveColumn(matrix, i));
       }
       return total;
   }
   
   public static int[][] RemoveColumn(int[][] matrix, int column)
   {
       int l = matrix[0].Length;
       int[][] newMatrix = new int[l - 1][];
       for(int i = 0; i < l - 1; i++)
           newMatrix[i] = new int[l - 1];
       
       for(int i = 1; i < l; i++)
       {
           int count = 0;
           for(int j = 0; j < l; j++)
           {
               if(j == column) continue;
               newMatrix[i - 1][count] = matrix[i][j];
               count++;
           }
       }
       return newMatrix;
   }
}

File: 4 kyu/Matrix Determinant/solution.ts
export function determinant(m: number[][]): number {
    if (m.length > 2) {
        return m[0].reduce(function (sum: number, val: number, idx: number): number {
            let add = (idx & 1) == 0 ? 1 : -1;
            let sub = m.slice(1).map(elem => elem.filter((_, i) => i != idx))
            return sum + add * val * determinant(sub);
        }, 0);
    } else if (m.length == 2) {
        return m[0][0] * m[1][1] - m[0][1] * m[1][0];
    } else {
        return m[0][0];
    }
}
_____________________________________________
export function determinant(m) {
  if (m.length === 1) {
    return m[0][0];
  } else {
    var res = 0;
    m[0].forEach((e, i) => {
      res += Math.pow(-1, i) * e * determinant(minor(m,i));
    });
    return res;
  }
}

function minor(m, i) {
  return m.slice(0).splice(1).map(e => e.filter((_, idx) => i != idx));
}
_____________________________________________
export const determinant = (m) => m.length === 1 ? 
  m[0][0] : 
  m[0].reduce((result, value, index) => result + Math.pow(-1, index) * value * determinant(minor(m, index)), 0);

const minor = (m, i) => m.slice(0).splice(1).map(e => e.filter((_, k) => i != k));
_____________________________________________
export function determinant(matrix: number[][]) {
  if (matrix.length === 1) {
    return matrix[0][0];
  }
  
  return matrix[0]
    .reduce(
      (d, e, index) => d + determinant(minor(matrix, index)) * e * sign(index),
      0
    );
}

function minor(matrix: number[][], target: number): number[][] {
  return removeFirstRow(matrix).map(row => removeColumn(row, target));
}

function removeFirstRow(matrix: number[][]): number[][] {
  return matrix.slice(1);
}

function removeColumn(row: number[], index: number): number[] {
  return [
    ...row.slice(0, index),
    ...row.slice(index + 1)
  ];
}

function sign(index: number): 1 | -1 {
  return index % 2 ? -1 : 1;
}
_____________________________________________
export function determinant(givenMatrix: number[][]): number {
    let determinantValue: number = 0;

    if (givenMatrix.length === 1) {
        determinantValue = givenMatrix[0][0];
    } else if (givenMatrix.length === 2) {
        determinantValue = (givenMatrix[0][0] * givenMatrix[1][1]) - (givenMatrix[1][0] * givenMatrix[0][1]);
    } else {
        const topValues: number[] = givenMatrix.splice(0, 1)[0];

        for (let i = 0; i < topValues.length; i++) {
            const numberSign: number = i % 2 === 0 ? 1 : -1;

            const newMatrix = givenMatrix.map(val => {
                return val.slice(0, i).concat(val.slice(i+1));
            });
            
            const currentValue = numberSign * (topValues[i] * determinant(newMatrix));

            determinantValue += currentValue;
        }
    }

    return determinantValue;
}
